keps/sig-node/4817-resource-claim-device-status/README.md

```diff
@@ -0,0 +1,727 @@
+interface name.
+
+`Data` being a slice of arbitrary data allows the DRA Driver to store
+device specific data in different formats. For example, a Network Device being
+configured by via a CNI plugin could get its `Data` field filled with the CNI
+result for troubleshooting purpose and with a Network result in a modern
+standard format (closer to Pod.Status.PodIPs for example) used by 3rd party
+controllers.
+
+For each device, if required, the DRA Driver processing the device allocation
+can report the status of it in the `Status.Devices` of the ResourceClaim
+by using the Kubernetes API.
+
+```golang
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the result of that was.
+type ResourceClaimStatus struct {
+    ...
+
+    // Devices contains the status of each device allocated for this
+    // claim, as reported by the driver. This can include driver-specific
+    // information. Entries are owned by their respective drivers.
+    //
+    // +optional
+    // +listType=map
+    // +listMapKey=driver
+    // +listMapKey=device
+    // +listMapKey=pool
+	  // +featureGate=DRAResourceClaimDeviceStatus
+    Devices []AllocatedDeviceStatus `json:"devices,omitempty" protobuf:"bytes,4,opt,name=devices"`
+}
+
+// AllocatedDeviceStatus contains the status of an allocated device, if the
+// driver chooses to report it. This may include driver-specific information.
+type AllocatedDeviceStatus struct {
+    // Driver specifies the name of the DRA driver whose kubelet
+    // plugin should be invoked to process the allocation once the claim is
+    // needed on a node.
+    //
+    // Must be a DNS subdomain and should end with a DNS domain owned by the
+    // vendor of the driver.
+    //
+    // +required
+    Driver string `json:"driver" protobuf:"bytes,1,rep,name=driver"`
+
+    // This name together with the driver name and the device name field
+    // identify which device was allocated (`<driver name>/<pool name>/<device name>`).
+    //
+    // Must not be longer than 253 characters and may contain one or more
+    // DNS sub-domains separated by slashes.
+    //
+    // +required
+    Pool string `json:"pool" protobuf:"bytes,2,rep,name=pool"`
+
+    // Device references one device instance via its name in the driver's
+    // resource pool. It must be a DNS label.
+    //
+    // +required
+    Device string `json:"device" protobuf:"bytes,3,rep,name=device"`
+
+    // Conditions contains the latest observation of the device's state.
+    // If the device has been configured according to the class and claim
+    // config references, the `Ready` condition should be True.
+    //
+    // +optional
+    // +listType=atomic
+    Conditions []metav1.Condition `json:"conditions" protobuf:"bytes,4,opt,name=conditions"`
+
+    // Data contains arbitrary driver-specific data.
+    //
+    // +optional
+    // +listType=atomic
+    Data []runtime.RawExtension `json:"data,omitempty" protobuf:"bytes,5,opt,name=data"`
```

Comment:
do we need this to be an array? 
```
grep -r \\[\\]runtime.RawExtension staging/src/k8s.io/api
staging/src/k8s.io/api/core/v1/zz_generated.deepcopy.go:                *out = make([]runtime.RawExtension, len(*in))
```
is a rawextension, so it can use  complex structures inside,  let's make this is a singular or is there any special motivation for making it an array?

Reply:
The main motivation was to report different types of data, like the [cni result](https://github.com/containernetworking/cni/blob/v1.2.3/pkg/types/100/types.go#L90), and the [multus network status](https://github.com/k8snetworkplumbingwg/network-attachment-definition-client/blob/v1.7.1/pkg/apis/k8s.cni.cncf.io/v1/types.go#L103).

cc @mskrocki @dougbtv

Reply:
I don't think we need to leak those details into the API, we are standardising the common fields, if some implementation want to use custom data  it has the RawExtension and model it the way they want, IIUIC it is perfectly fine to add those both types in the same Data field, and also we avoid having two actors on the same resource.
I'm not in favor of an array here and, in the worst scenario, if we find out that is needed, pluralising the field is not a breaking change,

Reply:
I changed it to a single RawExtension as suggested
```golang
type AllocatedDeviceStatus struct {
    ...
    // Data contains arbitrary driver-specific data.
    //
    // +optional
    Data runtime.RawExtension `json:"data,omitempty" protobuf:"bytes,5,opt,name=data"`
    ...
}
```